File: waterui-ffi/src/component/layout/scroll.rs
use waterui::component::layout::scroll::{Axis, ScrollView};

use crate::{waterui_anyview, IntoFFI};

#[repr(C)]
pub enum waterui_axis {
    HORIZONTAL,
    VERTICAL,
    ALL,
}

impl IntoFFI for Axis {
    type FFI = waterui_axis;
    fn into_ffi(self) -> Self::FFI {
        match self {
            Axis::Horizontal => waterui_axis::HORIZONTAL,
            Axis::Vertical => waterui_axis::VERTICAL,
            Axis::All => waterui_axis::ALL,
        }
    }
}

#[repr(C)]
pub struct waterui_scroll {
    content: *mut waterui_anyview,
    axis: waterui_axis,
}

into_ffi!(ScrollView, waterui_scroll, content, axis);

ffi_view!(
    ScrollView,
    waterui_scroll,
    waterui_view_force_as_scroll,
    waterui_view_scroll_id
);

File: waterui-swift/Sources/WaterUI/Scroll.swift
//
//  Scroll.swift
//  waterui-swift
//
//  Created by Lexo Liu on 10/20/24.
//
import SwiftUI
import CWaterUI

struct ScrollView:View,Component{
    static var id=waterui_view_scroll_id()
    var axis:SwiftUI.Axis.Set
    var content:WaterUI.AnyView
    var body: some View{
        SwiftUI.ScrollView(axis, content: {
            content
        })
    }
    
    init(anyview: OpaquePointer, env: Environment) {
        self.init(scroll: waterui_view_force_as_scroll(anyview), env: env)
    }
    
    init(scroll: waterui_scroll, env: Environment) {
        switch scroll.axis{
        case WATERUI_AXIS_VERTICAL:
            self.axis=[.vertical]
        case WATERUI_AXIS_HORIZONTAL:
            self.axis=[.horizontal]
            
        default:
            self.axis=[.vertical,.horizontal]
        }
        
        self.content=AnyView(anyview: scroll.content, env: env)
    }
}
